README.md
# <img src="https://uploads-ssl.webflow.com/5ea5d3315186cf5ec60c3ee4/5edf1c94ce4c859f2b188094_logo.svg" alt="Pip.Services Logo" width="200"> <br/> ElasticSearch components for Pip.Services in Java

This module is a part of the [Pip.Services](http://pipservices.org) polyglot microservices toolkit.

The Elasticsearch module contains logging components with data storage on the Elasticsearch server.

The module contains the following packages:
- **Build** - contains a factory for the construction of components
- **Log** - Logging components

<a name="links"></a> Quick links:

* [Configuration](http://docs.pipservices.org/toolkit/getting_started/configurations/)
* [Virtual memory configuration](https://www.elastic.co/guide/en/elasticsearch/reference/current/vm-max-map-count.html)
* [API Reference](https://pip-services3-java.github.io/pip-services3-elasticsearch-java/)
* [Change Log](CHANGELOG.md)
* [Get Help](http://docs.pipservices.org/get_help/)
* [Contribute](http://docs.pipservices.org/contribute/)

## Use

Go to the pom.xml file in Maven project and add dependencies::
```xml
<dependency>
  <groupId>org.pipservices3</groupId>
  <artifactId>pip-services3-elasticsearch</artifactId>
  <version>3.0.0</version>
</dependency>
```

Microservice components shall perform logging usual way using CompositeLogger component.
The CompositeLogger will get ElasticSearchLogger from references and will redirect log messages
there among other destinations.

```java
import org.pipservices3.commons.config.ConfigParams;
import org.pipservices3.commons.config.IConfigurable;
import org.pipservices3.commons.refer.IReferenceable;
import org.pipservices3.commons.refer.IReferences;
import org.pipservices3.components.log.CompositeLogger;

public class MyComponent implements IConfigurable, IReferenceable {
    private CompositeLogger _logger = new CompositeLogger();

    public void configure(ConfigParams config) {
        this._logger.configure(config);
    }

    public void  setReferences(IReferences refs) {
        this._logger.setReferences(refs);
    }

    public void myMethod(String correlationId, Object param1) {
        this._logger.trace(correlationId, "Executed method mycomponent.mymethod");
        // ....
    }
}
```

Configuration for your microservice that includes ElasticSearch logger may look the following way.

```yaml
...
{{#if ELASTICSEARCH_ENABLED}}
- descriptor: pip-services:logger:elasticsearch:default:1.0
  connection:
    uri: {{{ELASTICSEARCG_SERVICE_URI}}}
    host: {{{ELASTICSEARCH_SERVICE_HOST}}}{{#unless ELASTICSEARCH_SERVICE_HOST}}localhost{{/unless}}
    port: {{ELASTICSEARCG_SERVICE_PORT}}{{#unless ELASTICSEARCH_SERVICE_PORT}}9200{{/unless}}\ 
{{/if}}
...
```

## Develop

For development you shall install the following prerequisites:
* Java SE Development Kit 18+
* Eclipse Java Photon or another IDE of your choice
* Docker
* Apache Maven

Build the project:
```bash
mvn install
```

Run automated tests:
```bash
mvn test
```

Generate API documentation:
```bash
./docgen.ps1
```

Before committing changes run dockerized build and test as:
```bash
./build.ps1
./test.ps1
./clear.ps1
```

## Contacts

The initial implementation is done by
**Sergey Seroukhov**
**Danil Prisiazhnyi**


CHANGELOG.md
# <img src="https://uploads-ssl.webflow.com/5ea5d3315186cf5ec60c3ee4/5edf1c94ce4c859f2b188094_logo.svg" alt="Pip.Services Logo" width="200"> <br/> ElasticSearch components for Pip.Services in Java Changelog

## <a name="3.0.0"></a> 3.0.0 (2022-06-20)

Initial public release

### Features
* Added DefaultElasticSearchFactory
* Added ElasticSearchLogger

test/org/pipservices3/elasticsearch/fixtures/LoggerFixture.java
package org.pipservices3.elasticsearch.fixtures;

import org.junit.Test;
import org.pipservices3.components.log.CachedLogger;
import org.pipservices3.components.log.LogLevel;

import static org.junit.Assert.assertNotNull;
import static org.junit.Assert.assertTrue;

public class LoggerFixture {
    private final CachedLogger _logger;

    public LoggerFixture(CachedLogger logger) {
        this._logger = logger;
    }

    public void testLogLevel() {
        assertTrue(this._logger.getLevel().getValue() >= LogLevel.None.getValue());
        assertTrue(this._logger.getLevel().getValue() <= LogLevel.Trace.getValue());
    }

    public void testSimpleLogging() throws InterruptedException {
        this._logger.setLevel(LogLevel.Trace);

        this._logger.fatal("", "", "Fatal error message");
        this._logger.error("", "", "Error message");
        this._logger.warn(null, "Warning message");
        this._logger.info(null, "Information message");
        this._logger.debug(null, "Debug message");
        this._logger.trace(null, "Trace message");

        this._logger.dump();

        Thread.sleep(1000);
    }

    public void testErrorLogging() throws InterruptedException {
        Exception err;
        try {
            // Raise an exception
            throw new Exception();
        } catch (Exception ex) {
            err = ex;
            this._logger.fatal("123", ex, "Fatal error");
            this._logger.error("123", ex, "Recoverable error");
        }

        assertNotNull(err);

        this._logger.dump();

        Thread.sleep(1000);
    }
}


test/org/pipservices3/elasticsearch/log/ElasticSearchLoggerTest.java
package org.pipservices3.elasticsearch.log;

import org.junit.After;
import org.junit.Before;
import org.junit.Test;
import org.pipservices3.commons.config.ConfigParams;
import org.pipservices3.commons.errors.ApplicationException;
import org.pipservices3.commons.errors.InvocationException;
import org.pipservices3.elasticsearch.fixtures.LoggerFixture;

import static org.junit.Assert.assertTrue;

public class ElasticSearchLoggerTest {
    ElasticSearchLogger _logger;
    LoggerFixture _fixture;

    @Before
    public void setupClass() throws ApplicationException {
        var host = System.getenv("ELASTICSEARCH_SERVICE_HOST") != null ? System.getenv("ELASTICSEARCH_SERVICE_HOST") : "localhost";
        var port = System.getenv("ELASTICSEARCH_SERVICE_PORT") != null ? System.getenv("ELASTICSEARCH_SERVICE_PORT") : 9200;
        String dateFormat = "YYYYMMDD";

        _logger = new ElasticSearchLogger();
        _fixture = new LoggerFixture(_logger);

        var config = ConfigParams.fromTuples(
                "source", "test",
                "index", "log",
                "daily", true,
                "date_format", dateFormat,
                "connection.host", host,
                "connection.port", port
        );
        _logger.configure(config);

        _logger.open(null);
    }

    @After
    public void teardown() throws InvocationException {
        _logger.close(null);
    }

    @Test
    public void testLogLevel() {
        _fixture.testLogLevel();
    }

    @Test
    public void testSimpleLogging() throws InterruptedException {
        _fixture.testSimpleLogging();
    }

    @Test
    public void testErrorLogging() throws InterruptedException {
        _fixture.testErrorLogging();
    }

    /**
     * We test to ensure that the date pattern does not effect the opening of the elasticsearch component
     */
    @Test
    public void testDatePatternTestingYYYYMMDD() throws ApplicationException {
        var host = System.getenv("ELASTICSEARCH_SERVICE_HOST") != null ? System.getenv("ELASTICSEARCH_SERVICE_HOST") : "localhost";
        var port = System.getenv("ELASTICSEARCH_SERVICE_PORT") != null ? System.getenv("ELASTICSEARCH_SERVICE_PORT") : 9200;

        var logger = new ElasticSearchLogger();
        var dateFormat = "YYYY.MM.DD";

        var config = ConfigParams.fromTuples(
                "source", "test",
                "index", "log",
                "daily", true,
                "date_format", dateFormat,
                "connection.host", host,
                "connection.port", port
        );
        logger.configure(config);

        logger.open(null);

        // Since the currentIndex property is private, we will just check for an open connection
        assertTrue(logger.isOpen());
        logger.close(null);
    }

    /**
     * We test to ensure that the date pattern does not effect the opening of the elasticsearch component
     */
    @Test
    public void testDatePatternTestingYYYYMDD() throws ApplicationException {
        var host = System.getenv("ELASTICSEARCH_SERVICE_HOST") != null ? System.getenv("ELASTICSEARCH_SERVICE_HOST") : "localhost";
        var port = System.getenv("ELASTICSEARCH_SERVICE_PORT") != null ? System.getenv("ELASTICSEARCH_SERVICE_PORT") : 9200;
        String dateFormat = "YYYY.M.DD";

        var logger = new ElasticSearchLogger();

        var config = ConfigParams.fromTuples(
                "source", "test",
                "index", "log",
                "daily", true,
                "date_format", dateFormat,
                "connection.host", host,
                "connection.port", port
        );

        logger.configure(config);

        logger.open(null);

        // Since the currentIndex property is private, we will just check for an open connection
        assertTrue(logger.isOpen());
        logger.close(null);
    }
}


src/org/pipservices3/elasticsearch/log/ElasticSearchLogger.java
package org.pipservices3.elasticsearch.log;

import com.fasterxml.jackson.core.JsonProcessingException;
import org.apache.http.HttpHost;

import org.elasticsearch.action.admin.indices.get.GetIndexRequest;
import org.elasticsearch.action.bulk.BulkRequest;
import org.elasticsearch.action.index.IndexRequest;
import org.elasticsearch.client.*;
import org.elasticsearch.client.indices.CreateIndexRequest;

import org.elasticsearch.common.settings.Settings;
import org.pipservices3.commons.config.ConfigParams;
import org.pipservices3.commons.convert.JsonConverter;
import org.pipservices3.commons.data.IdGenerator;
import org.pipservices3.commons.errors.ApplicationException;
import org.pipservices3.commons.errors.ConfigException;
import org.pipservices3.commons.errors.InvocationException;
import org.pipservices3.commons.refer.IReferenceable;
import org.pipservices3.commons.refer.IReferences;
import org.pipservices3.commons.run.IOpenable;
import org.pipservices3.components.log.CachedLogger;
import org.pipservices3.components.log.LogMessage;
import org.pipservices3.rpc.connect.HttpConnectionResolver;


import java.io.IOException;
import java.time.ZonedDateTime;
import java.time.format.DateTimeFormatter;
import java.util.*;

/**
 * Logger that dumps execution logs to ElasticSearch service.
 * <p>
 * ElasticSearch is a popular search index. It is often used
 * to store and index execution logs by itself or as a part of
 * ELK (ElasticSearch - Logstash - Kibana) stack.
 * <p>
 * Authentication is not supported in this version.
 * <p>
 * ### Configuration parameters ###
 *
 * <pre>
 * - level:             maximum log level to capture
 * - source:            source (context) name
 * - connection(s):
 *     - discovery_key:         (optional) a key to retrieve the connection from {@link org.pipservices3.components.connect.IDiscovery}
 *     - protocol:              connection protocol: http or https
 *     - host:                  host name or IP address
 *     - port:                  port number
 *     - uri:                   resource URI or connection string with all parameters in it
 * - options:
 *     - interval:          interval in milliseconds to save log messages (default: 10 seconds)
 *     - max_cache_size:    maximum number of messages stored in this cache (default: 100)
 *     - index:             ElasticSearch index name (default: "log")
 *     - date_format        The date format to use when creating the index name. Eg. log-YYYYMMDD (default: "YYYYMMdd"). See {@link DateTimeFormatter#ofPattern}
 *     - daily:             true to create a new index every day by adding date suffix to the index name (default: false)
 *     - reconnect:         reconnect timeout in milliseconds (default: 60 sec)
 *     - timeout:           invocation timeout in milliseconds (default: 30 sec)
 *     - max_retries:       maximum number of retries (default: 3)
 *     - index_message:     true to enable indexing for message object (default: false)
 *     - include_type_name: Will create using a "typed" index compatible with ElasticSearch 6.x (default: false)
 * </pre>
 * <p>
 * ### References ###
 * <p>
 * - *:context-info:\*:*:1.0      (optional) {@link org.pipservices3.components.info.ContextInfo} to detect the context id and specify counters source
 * - *:discovery:*:*:1.0<         (optional) {@link org.pipservices3.components.connect.IDiscovery} services to resolve connection
 * <p>
 * ### Example ###
 * <pre>
 * {@code
 *     var logger = new ElasticSearchLogger();
 *     logger.configure(ConfigParams.fromTuples(
 *             "connection.protocol", "http",
 *             "connection.host", "localhost",
 *             "connection.port", 9200
 *     ));
 *
 *     logger.open("123");
 *
 *     var ex = new Exception();
 *     logger.error("123", ex, "Error occured: %s", ex.getMessage());
 *     logger.debug("123", "Everything is OK.");
 * }
 * </pre>
 */
public class ElasticSearchLogger extends CachedLogger implements IReferenceable, IOpenable {

    private final HttpConnectionResolver _connectionResolver = new HttpConnectionResolver();

    private Timer _timer;
    private String _index = "log";
    private String _dateFormat = "YYYYMMDD";
    private boolean _dailyIndex = false;
    private String _currentIndex;
    private int _reconnect = 60000;
    private int _timeout = 30000;
    private int _maxRetries = 3;
    private boolean _indexMessage = false;
    private boolean _include_type_name = false;
    private RestHighLevelClient _client = null;

    /**
     * Configures component by passing configuration parameters.
     *
     * @param config configuration parameters to be set.
     */
    @Override
    public void configure(ConfigParams config) {
        super.configure(config);

        this._connectionResolver.configure(config);

        this._index = config.getAsStringWithDefault("index", this._index);
        this._dateFormat = config.getAsStringWithDefault("date_format", this._dateFormat);
        this._dailyIndex = config.getAsBooleanWithDefault("daily", this._dailyIndex);
        this._reconnect = config.getAsIntegerWithDefault("options.reconnect", this._reconnect);
        this._timeout = config.getAsIntegerWithDefault("options.timeout", this._timeout);
        this._maxRetries = config.getAsIntegerWithDefault("options.max_retries", this._maxRetries);
        this._indexMessage = config.getAsBooleanWithDefault("options.index_message", this._indexMessage);
        this._include_type_name = config.getAsBooleanWithDefault("options.include_type_name", this._include_type_name);

        if (_dateFormat.contains("DD"))
            _dateFormat = _dateFormat.replace("DD", "dd");
    }

    /**
     * Sets references to dependent components.
     *
     * @param references references to locate the component dependencies.
     */
    @Override
    public void setReferences(IReferences references) {
        super.setReferences(references);
        this._connectionResolver.setReferences(references);
    }

    /**
     * Checks if the component is opened.
     *
     * @return true if the component has been opened and false otherwise.
     */
    @Override
    public boolean isOpen() {
        return this._timer != null;
    }

    /**
     * Opens the component.
     *
     * @param correlationId (optional) transaction id to trace execution through call chain.
     */
    @Override
    public void open(String correlationId) throws ApplicationException {
        if (this.isOpen())
            return;


        var connection = this._connectionResolver.resolve(correlationId);
        if (connection == null) {
            throw new ConfigException(
                    correlationId,
                    "NO_CONNECTION",
                    "Connection is not configured"
            );
        }

//        RestClientBuilder builder = RestClient.builder(
//                new HttpHost(connection.getHost(), connection.getPort(), "http"));

        RestClientBuilder builder = RestClient.builder(new HttpHost(connection.getHost(), connection.getPort(), "http"));

        builder.setRequestConfigCallback(
                builder1 -> builder1
                        .setConnectTimeout(_timeout)
                        .setSocketTimeout(_reconnect)
                        .setMaxRedirects(_maxRetries));

        _client = new RestHighLevelClient(builder);

        this.createIndexIfNeeded(correlationId, true);

        _timer = new Timer();

        _timer.scheduleAtFixedRate(new TimerTask() {
            @Override
            public void run() {
                dump();
            }
        }, 0, _interval);
    }

    /**
     * Closes component and frees used resources.
     *
     * @param correlationId (optional) transaction id to trace execution through call chain.
     */
    @Override
    public void close(String correlationId) throws InvocationException {
        this.save(this._cache);

        if (this._timer != null)
            this._timer.cancel();

        this._cache = new ArrayList<>();
        this._timer = null;
        this._client = null;
    }

    private String getCurrentIndex() {
        if (!this._dailyIndex) return this._index;

        var datePattern = ZonedDateTime.now().format(DateTimeFormatter.ofPattern(_dateFormat));

        return this._index + "-" + datePattern;
    }

    private void createIndexIfNeeded(String correlationId, boolean force) {
        var newIndex = this.getCurrentIndex();
        if (!force && Objects.equals(this._currentIndex, newIndex)) {
            return;
        }

        this._currentIndex = newIndex;

        GetIndexRequest request = new GetIndexRequest();
        request.indices(_currentIndex);

        boolean exists;

        try {
            exists = _client.indices().exists(request, RequestOptions.DEFAULT);
        } catch (IOException e) {
            throw new RuntimeException(e);
        }

        try {
            if (exists)
                return;

            var createRequest = new CreateIndexRequest(_currentIndex);

            createRequest.settings(Settings.builder()
                    .put("index.number_of_shards", 1)
            );

            createRequest.mapping(this.getIndexSchema());

            _client.indices().create(
                    createRequest, RequestOptions.DEFAULT
            );
        } catch (IOException err) {
            // Skip already exist errors
            if (!err.getMessage().contains("resource_already_exists"))
                throw new RuntimeException(err);
        }
    }

    /**
     * Returns the schema of the log message
     */
    private Map<String, ?> getIndexSchema() {
        final Map<String, ?> schema = Map.of(
                "properties", Map.of(
                        "time", Map.of("type", "date", "index", true),
                        "source", Map.of("type", "keyword", "index", true),
                        "level", Map.of("type", "keyword", "index", true),
                        "correlation_id", Map.of("type", "text", "index", true),
                        "error", Map.of(
                                "type", "object",
                                "properties", Map.of(
                                        "type", Map.of("type", "keyword", "index", true),
                                        "category", Map.of("type", "keyword", "index", true),
                                        "status", Map.of("type", "integer", "index", false),
                                        "code", Map.of("type", "keyword", "index", true),
                                        "message", Map.of("type", "text", "index", false),
                                        "details", Map.of("type", "object"),
                                        "correlation_id", Map.of("type", "text", "index", false),
                                        "cause", Map.of("type", "text", "index", false),
                                        "stack_trace", Map.of("type", "text", "index", false)
                                )
                        ),
                        "message", Map.of("type", "text", "index", this._indexMessage)
                )
        );

        if (this._include_type_name) {
            return Map.of("log_message", schema);
        } else return schema;
    }

    /**
     * Saves log messages from the cache.
     *
     * @param messages a list with log messages
     */
    @Override
    protected void save(List<LogMessage> messages) throws InvocationException {
        if (!this.isOpen() || messages.size() == 0)
            return;

        this.createIndexIfNeeded("elasticsearch_logger", false);

        var bulkRequest = new BulkRequest();
        for (var message : messages) {
            var req = getLogItem();

            try {
                req.source(JsonConverter.toMap(JsonConverter.toJson(message)));
                bulkRequest.add(req);
            } catch (JsonProcessingException e) {
                throw new RuntimeException(e);
            }
        }

        try {
            _client.bulk(bulkRequest, RequestOptions.DEFAULT);
        } catch (IOException e) {
            throw new RuntimeException(e);
        }
    }

    protected IndexRequest getLogItem() {
        if (this._include_type_name)
            return new IndexRequest(_currentIndex, "log_message", IdGenerator.nextLong()); // ElasticSearch 6.x
        else
            return new IndexRequest(_currentIndex).id(IdGenerator.nextLong()); // ElasticSearch 7.x
    }
}
